packages/gatsby/src/commands/build-javascript.ts

```diff
@@ -23,7 +26,44 @@ export const buildProductionBundle = async (
   )
 
   return new Promise((resolve, reject) => {
-    webpack(compilerConfig).run((err, stats) => {
+    const compiler = webpack(compilerConfig)
+
+    compiler.hooks.compilation.tap(`webpack-dep-tree-plugin`, compilation => {
+      compilation.hooks.seal.tap(`webpack-dep-tree-plugin`, () => {
+        const state = store.getState()
+        const mapOfTemplatesToStaticQueryHashes = mapTemplatesToStaticQueryHashes(
+          state,
+          compilation
+        )
+
+        mapOfTemplatesToStaticQueryHashes.forEach(
+          (staticQueryHashes, componentPath) => {
+            if (
+              !isEqual(
+                state.staticQueriesByTemplate.get(componentPath),
+                staticQueryHashes.map(String)
+              )
+            ) {
+              store.dispatch({
+                type: `ADD_PENDING_TEMPLATE_DATA_WRITE`,
+                payload: {
+                  componentPath,
+                },
+              })
+              store.dispatch({
+                type: `SET_STATIC_QUERIES_BY_TEMPLATE`,
+                payload: {
+                  componentPath,
+                  staticQueryHashes,
+                },
+              })
+            }
+          }
+        )
+      })
+    })
```

Comment:
This should be a webpack plugin instead of hacking it inside build-javascript command but this can be done later in the chain :)